contracts/staking/ERC20/StakingERC20.sol

```diff
@@ -1,97 +1,263 @@
+            staker.lastTimestampLocked = block.timestamp;
+        }
 
-        uint256 balanceAfter = IERC20(stakingToken).balanceOf(address(this));
-        uint256 amountTransferred = balanceAfter - balanceBefore;
+        // Do not transfer 0 rewards
+        if (rewards == 0) revert ZeroRewards();
 
-        staker.amountStaked += amountTransferred;
-        staker.lastUpdatedTimestamp = block.timestamp;
+        if (_getContractRewardsBalance() < rewards) revert InsufficientContractBalance();
 
-        totalStaked += amountTransferred;
+        // Because we update timestamps before transfer, any reentrancy attempt
+        // will use the current timestamps and calculate to 0
+        rewardsToken.safeTransfer(msg.sender, rewards);
 
-        emit Staked(msg.sender, amount, amountTransferred, stakingToken);
+        emit Claimed(msg.sender, rewards, address(rewardsToken));
     }
 
     /**
-     * @notice Unstake some or all of a user's stake
+     * @notice Unstake a specified amount of a user's non-locked stake
+     * 
      * @param amount The amount to withdraw
-     * @param exit If true, the user will unstake without claiming rewards (optional)
      */
-    function unstake(uint256 amount, bool exit) external override nonReentrant {
-        if (amount == 0) revert ZeroUnstake();
+    function unstake(uint256 amount) external override {
+        _unstake(amount, false, false);
+    }
+
+    /**
+     * @notice Unstake a specified amount of a user's locked funds that were locked
+     * @dev Will revert if funds are still within their lock period and not calling with `exit` as `true`
+     * 
+     * @param amount The amount to withdraw 
+     * @param exit Boolean if user wants to forfeit rewards
+     */
+    function unstakeLocked(uint256 amount, bool exit) public override {
+        _unstake(amount, true, exit);
+    }
+
+    /**
+     * @notice Return the time in seconds remaining for the staker's lock duration
+     */
+    function getRemainingLockTime() public view override returns (uint256) {
+        return _getRemainingLockTime(stakers[msg.sender]);
+    }
+
+    function getPendingRewards() public view override returns (uint256) {
+        return _getPendingRewards(stakers[msg.sender]);
+    }
+
+    function _stake(uint256 amount, uint256 lockDuration) internal {
+        // TODO be sure when new locks come in after an old lock is expired
+        // it is treated like a brand new lock
+        if (amount == 0) {
+            revert ZeroValue();
+        }
 
         Staker storage staker = stakers[msg.sender];
 
-        if (amount > staker.amountStaked) revert UnstakeMoreThanStake();
+        if (lockDuration == 0) {
+            // Get rewards they are owed from past unlocked stakes, if any
+            // `_getPendingRewards` will return 0 if `amountStaked == 0`
+            uint256 interimRewards =  _getInterimRewards(staker, block.timestamp - staker.lastTimestamp, false);
```

Comment:
second argument here seems redundant since we already pass the `staker` struct to the function. the time can be calced there without using extra calldata here